src/cMakeTaskProvider.ts

```diff
@@ -0,0 +1,179 @@
+        const task: CMakeTask = new Task(definition, scope, definition.label, CMakeTaskProvider.CMakeSourceStr,
+            new CustomExecution(async (resolvedDefinition: TaskDefinition): Promise<Pseudoterminal> =>
+              new CustomBuildTaskTerminal(resolvedDefinition.command, resolvedDefinition.target, this.cmakeDriver)
+            ), []); // TODO: add problem matcher
+        return task;
+    }
+    return undefined;
+  }
+
+}
+
+class CustomBuildTaskTerminal implements Pseudoterminal {
+  private writeEmitter = new EventEmitter<string>();
+  private closeEmitter = new EventEmitter<number>();
+  public get onDidWrite(): Event<string> { return this.writeEmitter.event; }
+  public get onDidClose(): Event<number> { return this.closeEmitter.event; }
+  private endOfLine: string = "\r\n";
+
+  constructor(private command: string, private target: string | null, private cmakeDriver: CMakeDriver | undefined) {
+  }
+
+  async open(_initialDimensions: TerminalDimensions | undefined): Promise<void> {
+      // At this point we can start using the terminal.
+      this.writeEmitter.fire(localize("starting_build", "Starting build...") + this.endOfLine);
+      await this.doBuild();
+  }
+
+  close(): void {
+      // The terminal has been closed. Shutdown the build.
+  }
+
+  private async doBuild(): Promise<any> {
+    if (this.command !== "build") {
+      this.writeEmitter.fire(localize("not_a_build_command", "Not a recognized build command.") + this.endOfLine);
+    }
+    let buildCommand: BuildCommand | null;
+    let cmakePath: string = "CMake.EXE";
+    let args: string[] | undefined = [];
+
+    // TODO: change the target from "all" to the default received from cmakedriver.
+    const defaultBuildTarget: string = this.target ? this.target : "all";
+    if (this.cmakeDriver) {
+      buildCommand = await this.cmakeDriver.getCMakeBuildCommand(defaultBuildTarget);
+      if (buildCommand) {
+        cmakePath = buildCommand.command;
+        args = buildCommand.args;
+      }
+    }
+    // TODO: find the correct cwd
+    const cwd: string = path.dirname(cmakePath) ? path.dirname(cmakePath) : path.dirname(defaultBuildTarget);
+    const options: cp.ExecOptions | undefined = { cwd: cwd };
+    let activeCommand: string = cmakePath.includes(" ") ? ("\"" + cmakePath + "\"") : cmakePath;
+    if (args) {
+      args.forEach(value => {
+        if (value.includes(" ")) {
+          value = "\"" + value + "\"";
+        }
+        activeCommand = activeCommand + " " + value;
+      });
+    }
+
+    const splitWriteEmitter = (lines: string | Buffer) => {
+      for (const line of lines.toString().split(/\r?\n/g)) {
+        this.writeEmitter.fire(line + this.endOfLine);
+      }
+    };
+    this.writeEmitter.fire(activeCommand + this.endOfLine);
+    try {
+      const result: number = await new Promise<number>((resolve) => {
+        cp.exec(activeCommand, options, (_error, stdout, _stderr) => {
```

Comment:
>exec [](http://example.com/codeflow?start=11&length=4)

I think we should use `child_process.spawn` instead. [https://nodejs.org/api/child_process.html#child_process_child_process_spawn_command_args_options](https://nodejs.org/api/child_process.html#child_process_child_process_spawn_command_args_options)

Otherwise it is too long to wait for all of the stdout to be processed.  The user will be confused and think the build is hung.  With `spawn` you can register callbacks for stdout and stderr that will fire each time a line is written.